File: app/lib/src/data/matchfeed.dart
// matchfeed.dart — Dart port of worker/src/matchfeed.js. Raw CORE plays feed →
// canonical MatchFeed JSON map (schema/canonical.ts). SOCCER ONLY (capability
// hasMatchFeed). Pure; faithful port kept in lock-step with the JS — the
// matchfeed golden suite asserts byte parity.
//
// The core resource is touch-by-touch (every pass/tackle/throw-in) with
// TEAM-RELATIVE pitch coordinates (x 0 = own goal line, 100 = opponent goal
// line; y 0..100 across), passes/shots also carrying where the ball ended
// (fieldPosition2X/Y). Participants are $refs only — athleteId parses out of
// the ref and joins the summary lineups downstream. Paginated and APPEND-ONLY;
// api.dart merges pages before calling this.

import 'util.dart';

num? _numOr(dynamic v) => v is num ? v : null;

final _teamRefRe = RegExp(r'/teams/(\d+)');
final _athleteRefRe = RegExp(r'/athletes/(\d+)');

String? _teamIdFromRef(dynamic ref) => ref is String ? _teamRefRe.firstMatch(ref)?.group(1) : null;
String? _athleteIdFromRef(dynamic ref) => ref is String ? _athleteRefRe.firstMatch(ref)?.group(1) : null;

// raw = the merged core plays doc {count, items[]} (pages merged oldest-first).
// homeId/awayId = the competition's team ids — core plays tag their team as a
// $ref, resolved to a side here.
Map<String, dynamic> normalizeMatchFeed(Map raw, dynamic homeId, dynamic awayId) {
  final items = raw['items'] is List ? raw['items'] as List : const [];
  final home = homeId != null ? jsStr(homeId) : '';
  final away = awayId != null ? jsStr(awayId) : '';
  final plays = <Map<String, dynamic>>[];
  for (final p in items) {
    if (p == null || field(p, 'valid') == false) continue;
    final type = field(field(p, 'type'), 'text');
    if (!truthy(type)) continue;
    final tid = _teamIdFromRef(field(field(p, 'team'), r'$ref'));
    final side = tid != null && tid == home
        ? 'home'
        : tid != null && tid == away
            ? 'away'
            : null;
    plays.add(pickT({
      'id': field(p, 'id') != null ? jsStr(field(p, 'id')) : null,
      'type': type,
      'period': _numOr(field(field(p, 'period'), 'number')),
      'clock': field(field(p, 'clock'), 'displayValue'),
      'sec': _numOr(field(field(p, 'clock'), 'value')),
      'side': side,
      'athleteId': _athleteIdFromRef(field(field(first(field(p, 'participants')), 'athlete'), r'$ref')),
      'shortText': field(p, 'shortText'),
      'text': field(p, 'text'),
      'x': _numOr(field(p, 'fieldPositionX')),
      'y': _numOr(field(p, 'fieldPositionY')),
      'x2': _numOr(field(p, 'fieldPosition2X')),
      'y2': _numOr(field(p, 'fieldPosition2Y')),
      'scoring': field(p, 'scoringPlay') == true ? true : null,
    }, ['id', 'type', 'period', 'clock', 'sec', 'side', 'athleteId', 'shortText', 'text', 'x', 'y', 'x2', 'y2', 'scoring']));
  }
  return {'count': raw['count'] is num ? raw['count'] : plays.length, 'plays': plays};
}

File: worker/src/matchfeed.js
// matchfeed.js — raw CORE plays feed → canonical MatchFeed (see canonical.ts).
// SOCCER ONLY (capability hasMatchFeed). Pure function; no I/O.
//
// The core resource: /v2/sports/soccer/leagues/{lg}/events/{id}/competitions/{id}/plays
// VERIFIED 2026-07 (fifa.world, live): touch-by-touch — every pass, tackle,
// throw-in — each play carrying TEAM-RELATIVE pitch coordinates (fieldPositionX
// 0 = own goal line, 100 = opponent goal line; fieldPositionY 0..100 across),
// with passes/shots also carrying fieldPosition2X/Y (where the ball ended up).
// Participants are $refs only (no names) — athleteId is parsed from the ref and
// joined against the summary lineups downstream; shortText ('W. Saliba Pass')
// is the self-contained fallback label. Paginated ~25/page by default, ?limit
// honored to ≥300; APPEND-ONLY, so the caller caches full pages and merges.
//
// This deliberately is NOT the narrative feed (that's summary commentary[]) —
// it exists for the live-pitch view (pass trail, possession, restarts), the
// shot map (start→end trajectories) and the derived momentum chart.

const numOr = v => (typeof v === 'number' ? v : undefined);
const pick = (o, keys) => Object.fromEntries(keys.filter(k => o[k] != null && o[k] !== '').map(k => [k, o[k]]));

const teamIdFromRef = ref => {
  const m = typeof ref === 'string' ? ref.match(/\/teams\/(\d+)/) : null;
  return m ? m[1] : undefined;
};
const athleteIdFromRef = ref => {
  const m = typeof ref === 'string' ? ref.match(/\/athletes\/(\d+)/) : null;
  return m ? m[1] : undefined;
};

// raw = the merged core plays doc {count, items[]} (caller merges pages, in
// page order — ESPN pages oldest-first so the merge is chronological).
// homeId/awayId = the competition's team ids (from the scoreboard/summary
// header) — core plays tag their team as a $ref, resolved to a side here.
export function normalizeMatchFeed(raw, homeId, awayId) {
  const items = Array.isArray(raw?.items) ? raw.items : [];
  const home = homeId != null ? String(homeId) : '';
  const away = awayId != null ? String(awayId) : '';
  const plays = [];
  for (const p of items) {
    if (!p || p.valid === false) continue;
    const type = p.type?.text;
    if (!type) continue;
    const tid = teamIdFromRef(p.team?.$ref);
    const side = tid && tid === home ? 'home' : tid && tid === away ? 'away' : undefined;
    plays.push(pick({
      id: p.id != null ? String(p.id) : undefined,
      type,
      period: numOr(p.period?.number),
      clock: p.clock?.displayValue,
      sec: numOr(p.clock?.value),
      side,
      athleteId: athleteIdFromRef(p.participants?.[0]?.athlete?.$ref),
      shortText: p.shortText,
      text: p.text,
      x: numOr(p.fieldPositionX),
      y: numOr(p.fieldPositionY),
      x2: numOr(p.fieldPosition2X),
      y2: numOr(p.fieldPosition2Y),
      scoring: p.scoringPlay === true ? true : undefined,
    }, ['id', 'type', 'period', 'clock', 'sec', 'side', 'athleteId', 'shortText', 'text', 'x', 'y', 'x2', 'y2', 'scoring']));
  }
  return { count: typeof raw?.count === 'number' ? raw.count : plays.length, plays };
}
